File: packages/scan/src/stage-caching/option-fingerprint.mapper.ts
/**
 * A stable key for an option bag, so a stage knows whether it has already run
 * with these settings.
 *
 * Not `JSON.stringify`: key order would make two identical bags disagree, an
 * explicit `undefined` would differ from an absent key, an OCR engine or a
 * normalisation predicate would throw or serialise to nothing, and a raster
 * would stringify megabytes of pixels to decide a cache hit.
 *
 * So keys are sorted, `undefined` is elided, and anything that cannot be
 * compared by value - a function, an engine, a typed array - is compared by
 * **identity**, through a counter kept in a `WeakMap`. Two calls passing the
 * same engine agree; two calls passing equivalent-looking engines do not, which
 * is the safe direction: a false miss costs time, a false hit returns the wrong
 * answer.
 */

const identities = new WeakMap<object, number>()
let nextIdentity = 0

function identity (value: object): number {
  let id = identities.get(value)
  if (id === undefined) {
    id = nextIdentity++
    identities.set(value, id)
  }

  return id
}

function stable (value: unknown): unknown {
  if (value === null || typeof value === 'number' || typeof value === 'string' || typeof value === 'boolean') return value
  if (typeof value === 'bigint') return `bigint:${value}`
  if (typeof value === 'function') return `fn#${identity(value)}`
  if (typeof value !== 'object') return String(value)

  if (ArrayBuffer.isView(value)) return `bytes:${(value).byteLength}#${identity(value)}`
  if (value instanceof ArrayBuffer) return `buffer:${value.byteLength}#${identity(value)}`
  if (value instanceof URL) return `url:${value.href}`
  if (Array.isArray(value)) return value.map(item => stable(item))

  const raster = value as { width?: unknown, height?: unknown, data?: unknown }
  // A raster is identified by its shape and which object it is, never by its pixels.
  if (typeof raster.width === 'number' && typeof raster.height === 'number' && ArrayBuffer.isView(raster.data))
    return `raster:${raster.width}x${raster.height}#${identity(value)}`

  const entries = Object.entries(value as Record<string, unknown>)
    .filter(([, of]) => of !== undefined)
    .toSorted(([a], [b]) => a.localeCompare(b))
    .map(([key, of]) => [key, stable(of)] as const)

  // A plain bag is what it holds, so `{ ocr: undefined }` is `{}` - the whole
  // point of eliding `undefined` is that an absent key and an explicit one are
  // the same settings. Anything else that yields no entries is opaque rather
  // than empty - a Map, a Set, a Date, a class instance keeping its state
  // privately - and two of those are the same only if they are the same object.
  const prototype = Object.getPrototypeOf(value) as unknown
  const plain = prototype === Object.prototype || prototype === null
  if (!plain && entries.length === 0) return `opaque#${identity(value)}`

  return Object.fromEntries(entries)
}

/** A stable string for any option bag. Equal strings mean equal settings. */
export function fingerprint (value: unknown): string {
  return JSON.stringify(stable(value)) ?? 'undefined'
}

File: python-packages/scanmate-scan/scanmate_scan/stage_caching/option_fingerprint_mapper.py
"""A stable key for an option bag, so a stage knows whether it has already run.

Not a plain ``json.dumps``: key order would make two identical bags disagree, an
absent setting would differ from one passed explicitly, an OCR engine or a
predicate cannot be serialised at all, and a raster would stringify megabytes of
pixels to decide a cache hit.

So keys are sorted, ``None`` is elided, and anything that cannot be compared by
value - a function, an engine, an array of pixels - is compared by **identity**,
through a counter keyed on ``id()``. Two calls passing the same engine agree;
two calls passing equivalent-looking engines do not, which is the safe
direction: a false miss costs time, a false hit returns the wrong answer.

.. note::
   **The string itself is not held to the TypeScript's**, and it is the one
   thing in this port that could not be even in principle. It carries identity
   counters, which are per-process by definition; the TypeScript sorts keys
   with ``localeCompare``, which is locale-aware rather than by code point; and
   a cache key is never handed between the two languages, so an identical
   spelling would buy nothing. What *is* held is the contract the key exists
   for: **equal settings give equal keys, and different settings give different
   ones.** That is what the tests check.
"""

from __future__ import annotations

import json
from typing import Any

#: Identity numbers handed out to values that cannot be compared by value.
#: Keyed on ``id()``, with the value kept alive alongside so the id cannot be
#: recycled onto a different object while the cache still holds the key -
#: which would be a false HIT, the one direction that must never happen.
_identities: dict[int, tuple[int, Any]] = {}


def _identity(value: object) -> int:
    """A stable number for one object, for as long as this process runs."""
    key = id(value)
    known = _identities.get(key)
    if known is not None:
        return known[0]
    number = len(_identities)
    _identities[key] = (number, value)

    return number


def _stable(value: object) -> object:
    """The value as something comparable, or a token standing for it."""
    if value is None or isinstance(value, (str, int, float)) and not isinstance(value, bool):
        return value
    if isinstance(value, bool):
        return value
    if callable(value):
        return f"fn#{_identity(value)}"
    if isinstance(value, (bytes, bytearray, memoryview)):
        return f"bytes:{len(bytes(value))}#{_identity(value)}"
    if isinstance(value, (list, tuple)):
        return [_stable(item) for item in value]

    # A raster or a greyscale image is identified by its shape and which object
    # it is, never by its pixels.
    pixels = getattr(value, "pixels", None)
    if pixels is not None and hasattr(pixels, "shape"):
        width = getattr(value, "width", None)
        height = getattr(value, "height", None)

        return f"raster:{width}x{height}#{_identity(value)}"

    fields = _fields_of(value)
    if fields is None:
        return f"opaque#{_identity(value)}"

    # `None` is elided, so a setting left at its default and one passed
    # explicitly as None are the same settings - which is the whole point.
    return {key: _stable(item) for key, item in sorted(fields.items()) if item is not None}


def _fields_of(value: object) -> dict[str, Any] | None:
    """The value's settings, or ``None`` when it keeps its state privately."""
    if isinstance(value, dict):
        return {str(key): item for key, item in value.items()}
    if hasattr(value, "__dataclass_fields__"):
        return {name: getattr(value, name) for name in value.__dataclass_fields__}

    return None


def fingerprint(value: object) -> str:
    """A stable string for any option bag.

    :param value: The settings.
    :returns: A string that is equal exactly when the settings are.
    """
    return json.dumps(_stable(value), sort_keys=True, separators=(",", ":"), default=str)
